File: backend/app/core/patient_resolver.py
"""Shared utility to resolve patient_id for both patient and caregiver users."""

from fastapi import Depends, HTTPException

from .database import get_db
from .security import get_current_user


async def resolve_patient_id(supabase_uid: str = Depends(get_current_user)) -> str:
    """
    Dependency that returns the patient_id for the current user.

    - If the user is a patient, returns their own patient doc ID.
    - If the user is a caregiver, returns their linked patient_id.
    - Raises 404 if no patient is linked.
    """
    db = get_db()
    user = await db["users"].find_one({"supabase_uid": supabase_uid})
    if not user:
        raise HTTPException(status_code=404, detail="Profile not found. Sign in again.")

    patient_id = user.get("patient_id")
    if not patient_id:
        raise HTTPException(
            status_code=404,
            detail="No patient linked to your account. "
            + (
                "Ask your patient for their link code."
                if user.get("role") == "caregiver"
                else "Account setup incomplete."
            ),
        )
    return str(patient_id)

File: src/server-core/patientResolver.ts
import { Request, Response, NextFunction } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "./database";

// Extend Express Request to include patientId
declare global {
  namespace Express {
    interface Request {
      patientId?: string;
    }
  }
}

export async function resolvePatientId(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  const db = getDb();
  const supabaseUid = req.auth?.userId;
  if (!supabaseUid) {
    res.status(401).json({ detail: "Not authenticated" });
    return;
  }

  const user = await db.collection("users").findOne({ supabase_uid: supabaseUid });
  if (!user) {
    res.status(404).json({ detail: "Profile not found. Sign in again." });
    return;
  }

  const patientId = user.patient_id;
  if (patientId && !ObjectId.isValid(String(patientId))) {
    res.status(500).json({ detail: "Account data corrupted. Please sign out and back in." });
    return;
  }
  if (!patientId) {
    const msg =
      user.role === "caregiver"
        ? "Ask your patient for their link code."
        : "Account setup incomplete.";
    res.status(404).json({ detail: `No patient linked to your account. ${msg}` });
    return;
  }

  req.patientId = String(patientId);
  next();
}
